2020/16/ticket/src/main.rs

```diff
@@ -0,0 +1,257 @@
+use std::collections::HashMap;
+
+fn main() -> Result<(), std::io::Error> {
+    let args: Vec<_> = std::env::args().collect();
+    let input = std::fs::read_to_string(&args[1])?;
+
+    let conditions = ConditionSet::parse(&input);
+    let tickets = Ticket::parse_nearby(&input);
+    if let Some(mine) = Ticket::parse_yours(&input) {
+        let res = conditions.order_fields(&tickets);
+        let mut product = 1;
+        for (name, _) in &conditions.fields {
+            if name.starts_with("departure") {
+                product *= mine.field_values[res[&name]];
+            }
+        }
+
+        println!("The product is: {}", product);
+    } else {
+        println!("Couldn't parse my ticket")
+    }
+
+    Ok(())
+}
+
+#[derive(Debug)]
+pub struct Range {
+    min: usize,
+    max: usize,
+}
+
+impl Range {
+    pub fn new(min: &str, max: &str) -> Self {
+        Range {
+            min: min.parse().unwrap(),
+            max: max.parse().unwrap(),
+        }
+    }
+
+    fn in_ranges(v: usize, r1: &Range, r2: &Range) -> bool {
+        v >= r1.min && v <= r1.max || v >= r2.min && v <= r2.max
+    }
+}
+
+pub struct ConditionSet {
+    fields: HashMap<String, (Range, Range)>,
```

Comment:
I wish Go had tuples instead of just janky "multiple return values"

Reply:
I've been missing C#'s named tuple elements though.